cmd/restic/cmd_cleanup.go

```diff
@@ -0,0 +1,372 @@
+			repackFreeBlobs += p.unusedBlobs
+		}
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+
+	// If there is only one small Pack to repack, there is nothing to do...
+	if len(repackSmallPacks) == 1 && len(repackPacks) == 0 {
+		repackSmallPacks = restic.NewIDSet()
+		repackBytes = 0
+		repackBlobs = 0
+		repackFreeBytes = 0
+		repackFreeBlobs = 0
+	}
+
+	// The size informations are approximations because:
+	// - it is not a priori known how many packs will result from the repacking
+	// - rebuilding the index will also free some space
+	Verbosef("found %d unreferenced packs\n", len(removePacksFirst))
+	Verbosef("found %d no longer used packs with %d blobs\n", len(removePacks), removeBlobs)
+	Verbosef("deleting unused packs will free about %s\n\n", formatBytes(removeBytes))
+
+	Verbosef("found %d partly used packs + %d small packs = total %d packs with %d blobs for repacking with %s\n",
+		len(repackPacks), len(repackSmallPacks), len(repackPacks)+len(repackSmallPacks), repackBlobs, formatBytes(repackBytes))
+	Verbosef("repacking packs will delete %d blobs and free about %s\n", repackFreeBlobs, formatBytes(repackFreeBytes))
+
+	Verbosef("cleanup-packs will totally delete %d blobs and free about %s\n", removeBlobs+repackFreeBlobs, formatBytes(removeBytes+repackFreeBytes))
+
+	// from now on, treat small packs like all other to-repack packs
+	repackPacks.Merge(repackSmallPacks)
+
+	// unreferenced packs can be saveley deleted first
+	if len(removePacksFirst) != 0 {
+		Verbosef("deleting unreferenced pack files...\n")
+		DeleteFiles(gopts, opts.DryRun, true, repo, removePacksFirst, restic.DataFile)
+	}
+
+	if len(repackPacks) > 0 {
+		if !opts.DryRun {
+			Verbosef("repacking packs...\n")
+			bar := newProgressMax(!gopts.Quiet, uint64(repackBlobs), "blobs repacked")
+			bar.Start()
+			// TODO in Repack:  - Parallelize repacking
+			//                  - Save full indexes during repacking
+			//                  - Make use of blobs stored multiple times (e.g. if SHA doesn't match)
+			obsolete, err := repository.Repack(ctx, repo, repackPacks, keepBlobs, bar)
+			if err != nil {
+				return err
+			}
+			bar.Done()
+			// all packs-to-repack should now be obsolete!
+			if !obsolete.Equals(repackPacks) {
+				return errors.New("error: obsolete packs do not match packs-to-repack!")
+			}
+		} else {
+			if !gopts.JSON && gopts.verbosity >= 2 {
+				for id := range repackPacks {
+					Verbosef("would have repacked pack %v.\n", id.Str())
+				}
+			}
+		}
+
+		// Also remove repacked pack at the end!
+		removePacks.Merge(repackPacks)
+	}
+
+	var obsoleteIndexes restic.IDSet
+	if len(removePacks) > 0 && !opts.DryRun {
+		Verbosef("updating index files...\n")
+
+		// TODO in RebuildIndex: - Save full indexes
+		//						- Parallelize repacking
+		//						- make it work with already saved indexes during Repack above
+		obsoleteIndexes, err = (repo.Index()).(*repository.MasterIndex).RebuildIndex(ctx, removePacks)
```

Comment:
I would also be an option to reuse the existing `rebuildIndex` command and defer the in-memory rebuild to later.

However, that would have two downsides: it would be a lot slower and have the ugly problem that if the header of some packs that are contained in the index are for some reason no longer readable, then `cleanup` would loose those packs from the index. For `prune` that's rather unlikely to happen as it rebuilds the index before making any modifications. Thus I think that the in-memory rebuild should be part of this PR, but some of the necessary groundwork should, for merging, end up in its own PR.

The in-memory index rebuild would also avoid the most ugly parts of #2685 that are necessary to properly handle eventually consistent (in the sense that it may take a short time until you can see newly written packs) backends.

Reply:
I changed this part and provide two options (both not optimal) to rebuild the index.